Directory walk routines

harendra-kumar opened the pull request:
I added a generic directory walk functionality which walks a dir tree and calls a handler during traversal. All other variants of dir recurse functionality can be implemented in terms of this one core function. It also avoids loops in traversal due to symlinks or hardlinks.

It may be better to review one commit at a time. First of all, please take a look if the APIs, naming etc. looks ok or if you have any better suggestions.

I also added a few more `listDirRecur` variants, if you think those are a bloat I am fine to leave them out.

mrkkrp commented:
Thanks, I'll review it in more detail later, but for now we need to fix the build for GHC 7.8 and add package version bounds for the new dependencies.

harendra-kumar commented:
All versions of `containers` and `unix-compat` have what I am using, so I left them unbounded. Is there any guideline other than the API availability and compatibility?

mrkkrp commented:
Things to change:

* Between docs and function there should be an empty line (this is how the library is formatted anyway, so let's be consistent).

* Don't end docs with one line containing only `--`.

* Unless I'm missing something `listDirRecurWithPruned` is not necessary as its functionality can be achieved with `listDirRecurWith` and `not` used with predicates.

* I sort of hate type synonyms unless they are really make life easier, let's remove `WalkHandler`, if I decide to use the `walkDir` function I would need to look up it every time, so let's keep things explicit here.

* Strings after `-- ^` should start with a capital letter (but no full stops).

* Let's call `walkDir'` `walkDirAccum` or something.

* Let's drop `walkDirAll` — it's trivial to achieve the same effect with `walkDir`. (And so also `walkDirAll'`).

* Update the `CHANGELOG.md` file accordingly. Write under `Path IO 0.1.2`, but don't bump version.

* Add tests for new functions. The existing tests may be helpful.

That's about it. Good work, I like the RP, looking forward to review the updated version.

harendra-kumar commented:
Thanks for quick and detailed review.

> Unless I'm missing something listDirRecurWithPruned is not necessary as its functionality can be achieved with listDirRecurWith and not used with predicates.

I think I need to make the documentation clearer then. The pruned version prints the directory but not its children. It is equivalent to the Unix `find -prune`. If we use a `not` predicate on dir then neither the dir nor the children will be printed.

> I sort of hate type synonyms unless they are really make life easier, let's remove WalkHandler, if I 

Even I do not like it. I made a compromise because of a limitation of haddock. With the synonym I can easily document the signature. But If I keep the signature in signature haddock does not allow documentation of inner signature. The alternative could be to document it outside instead along with the main documentation. We can also raise a haddock issue.

> Let's call walkDir' walkDirAccum or something.

I thought about names like `walkDirLog`, `walkDirCollect`, `walkDirWriter` and even `walkDirAccumulate` but did not like them much so I just kept a vague name which is perhaps worse. I do not have a strong opinion on this though, you decide a name and let me know. `walkDirAccum` then?

> Let's drop walkDirAll — it's trivial to achieve the same effect with walkDir. (And so also
Alright. They provide a simpler handler but on the flip side they might cause confusion because of one handler type.

> Add tests for new functions. The existing tests may be helpful.

`walkDir` gets tested via other functions though I can add a test for loop avoidance case. I will add tests for the other top level functions that we are retaining post review.

harendra-kumar commented:
How does it look now? I will be happy to address any suggestions. I will update the changelog and tests once the main content gets finalized.

mrkkrp commented:
Where [this](https://github.com/mrkkrp/path-io/pull/11/files#diff-06c7d82a14aee6aafd149f432cc2dc17R362) is supposed to be visible? It's not a part of Haddock for any public function. In this just internal notes?

I'm also thinking how efficient [this code](https://github.com/mrkkrp/path-io/pull/11/files#diff-06c7d82a14aee6aafd149f432cc2dc17R328) is going to be. My original implementation is also sub-optimal, maybe we should use difference lists to efficiently grow the resulting list of all files (which may be quite long). This is rather a note to myself, you don't have to change it.

[Here](https://github.com/mrkkrp/path-io/pull/11/files#diff-06c7d82a14aee6aafd149f432cc2dc17R406), how about:

```haskell
walkDir handler topdir = void (makeAbsolute topdir >>= walkAvoidLoop Set.empty)
```

Although it's not that important.

LGTM, you can start with changelog and tests, thanks.

harendra-kumar commented:
> Where this is supposed to be visible? It's not a part of Haddock for any public function. In this just

These are just implementation notes for developers. I wrote down what I was thinking when I was implementing it and kept it there for others to understand the reasoning behind it. 

Some perf experiments:
```
Listing a directory tree with ~100000 files

find:
real  0m23.791s
user  0m0.468s
sys 0m7.576s

listDirRecur (original):
real  1m12.409s
user  0m11.732s
sys 0m32.220s

walkDir (just traverse no printing):
real  1m15.644s
user  0m7.288s
sys 0m34.868s

walkDirAccum (print accumulated value):
real  1m17.612s
user  0m8.392s
sys 0m36.996s
```

`walkDirAccum` is around 7% slower compared to the original `listDirRecur`, I guess it could be mainly due to the loop avoidance logic. `find` is very fast, our goal should be to get close to that.

> walkDir handler topdir = void (makeAbsolute topdir >>= walkAvoidLoop Set.empty)

This is better I changed it.

harendra-kumar commented:
I noticed that the user+sys time in the old `listDirRecur` and `walkDirAccum` is not very much different (just a 1.5 sec diff). So much of the difference in total time is blocked time and not the processing time.

harendra-kumar commented:
Ah, for 7.8.4 use of `void` requires a `Functor` instance for `Monad m` wherever `walkDir` is called. All users will also have to do the same. I will revert it to the old way.

harendra-kumar commented:
Done. Take a look again. I may have to rebase to simplify the history because there are a couple of reverts.

harendra-kumar commented:
Do we have a windows CI setup?

mrkkrp commented:
No. We could setup it if it's free for open source, though.

harendra-kumar commented:
I changed the `WalkDescend` API to the inverse i.e. `WalkExclude`. Please see the commit message for justification. Let me know if this looks ok.

Another change that I am contemplating is to pass a list of handlers so that we can compose any combination of filters rather than proliferating more APIs for specific tasks.

mrkkrp commented:
OK, I'm a little busy right now, but I remember about this PR and I'll review and perhaps setup Appveyor when I have time.

harendra-kumar commented:
No worries and no hurry.

harendra-kumar commented:
The earlier perf figures of find were not correct because find does not follow symlinks by default and walkDir does. Here are new fair comparison figures:

```
walkDirAccum:
real  1m18.470s
user  0m10.716s
sys 0m37.412s

find -L
real  1m7.172s
user  0m0.264s
sys 0m32.676s
```

Its not bad after all.

harendra-kumar commented:
This should be the final commit in this PR except for, bug fixing, addressing any review comments or some changelog tweaking once the review is done. 

In this commit, I changed the `walkDirAccum` signature to accept two handlers for better modularity by separating the descend and output concerns. Now we can compose using predicates for descending or output writing. I also added some handy utility functions for descending and writing.

harendra-kumar commented:
Note on symlink following behavior: `walkDir` always requires you to specify a handler which decides how to descend and therefore there is no default behavior as such.

However `walkDirAccum` has a default now because I have made descend handler optional in the last commit. We have to decide what default we want to keep when the descend handler is not specified by the user. Another option is to make the descend handler mandatory in which case the user has to explicitly specify the behavior. Currently the default behavior is follow symlinks.

mrkkrp commented on Path/IO.hs:
```diff
@@ -1,3 +1,4 @@
+-- vim:ts=2 sw=2
```
Please remove this. I prefer not to keep editor-specific settings in source files.

harendra-kumar replied:
Sorry, this was not intentional.

mrkkrp commented on Path/IO.hs:
```diff
@@ -314,11 +329,187 @@ listDir path = do
+-- matching a predicate.
+
+descendWith :: (MonadIO m
+#if !MIN_VERSION_base(4,8,0)
```
Remove this and use `liftM` instead of `fmap`.

harendra-kumar replied:
Of course!

mrkkrp commented on Path/IO.hs:
```diff
@@ -314,11 +329,187 @@ listDir path = do
+  => (Path Abs Dir -> m Bool)                 -- ^ Directory match predicate
+  -> (Path Abs File -> m Bool)                -- ^ File match predicate
+  -> (Path Abs Dir -> [Path Abs Dir] -> [Path Abs File]
+      -> m ([Path Abs Dir], [Path Abs File])) -- ^ Output tuple writer
```
Please align it with previous `->` arrows.

harendra-kumar replied:
I indented that because it is a continuation of a grouped type signature. The grouping may not be evident without the indentation.

mrkkrp commented on Path/IO.hs:
```diff
@@ -314,11 +329,187 @@ listDir path = do
+
+----------------------------------------------------------------------------
+-- Directory Walk
+----------------------------------------------------------------------------
```
Please preserve style of headers. Every part in import section should have corresponding header in code with exactly the same header, looking like this, e.g.:

```haskell
----------------------------------------------------------------------------
-- Walking directory trees
```

The line of `-` is 76 characters long.

harendra-kumar replied:
Ah, I did not notice that. I reorganized the module exports a little bit. Should I move the function bodies too in the corresponding sections? I guess they should be under the correct header but not necessarily in the same order. I mean do not move just to reorder within the same section but move if in wrong section.

mrkkrp commented:
I left some more comments, mostly cosmetic things. Once you fix this, we can merge and release new version, I guess. Thanks for your contribution.

harendra-kumar commented:
I have made all changes except for the arrow alignment. Let me know your response on that. Also, let me know if you have any more suggestions or opinion on the API semantics, naming etc.

mrkkrp commented:
Oh, I missed that thing with the arrow. I'll take a look once again and merge once I have time.

mrkkrp commented:
Applied as  4f2cffb, thank you!
